### PR TITLE
Introduce integration testing framework

### PR DESCRIPTION
**Pull request by trestletech**
This PR introduces an integration testing framework that facilitates the testing of reactives inside of a Shiny server function or a Shiny module. The two entry-points are `testModule` and `testServer` which take a module or a shiny application object, respectively.

The vignette included in this PR best captures the functionality and how to use this. The initial version is published here: https://beta.rstudioconnect.com/content/6874/integration-testing.html

Intentionally omitted from this PR is any inline documentation on these new functions or their associated objects. I'm still debating whether this would be best served by an article in the dev center, something roxygenized, more vignettes, or some combination. I'll start work on how to document and expose this stuff next. ([Preview](https://github.com/rstudio/shiny/compare/jeff/int-test...jeff/int-test-docs?expand=1))

## Validation

This PR finally surfaces some of the internal work that we've been doing for a few weeks now. So there's a little more surface area that we could test here. However, any edge cases that we've discovered around weird reactivity or timing issues are already covered in unit tests, so I don't think there will be too much exploring required.

Take a look at `testServer` and `testModule` and pass in a few applications or modules as discussed in the vignette (linked above). If everything there works, I think we may be good for now. Once the documentation for all this lands, I'll want to get more feedback on usability and approachability, but that's intentionally omitted right now.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -10,3 +10,4 @@ shinyapps/
 README.html
 .*.Rnb.cached
 tools/yarn-error.log
+vignettes/*.R
```
Comment (wch) [inline .gitignore]: Why do these files need to be ignored?

File: R/test-module.R [hunk 2]
```diff
@@ -0,0 +1,117 @@
+
+
+#' Test a shiny module
+#' @param module The module under test
```
Comment (wch) [inline R/test-module.R]: "under test" is a little confusing -- "The module to test" seems clearer to me.

File: .gitignore (hunk 1 above)
Comment (trestletech) [reply .gitignore]: I have no idea. Generating vignettes seems to create a .R and a .html file in the vignettes directory. I cargo-culted this from dplyr (https://github.com/tidyverse/dplyr/blob/master/.gitignore#L12) and have just been manually ignoring/deleting the generated HTML file... 🙈

If you know better, I'm happy to revise.

File: R/test-module.R [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  # Capture the environment from the module
+  # Inserts `session$env <- environment()` at the top of the function
+  fn_body <- body(module)
+  fn_body[seq(3, length(fn_body)+1)] <- fn_body[seq(2, length(fn_body))]
+  fn_body[[2]] <- quote(session$env <- environment())
+  body(module) <- fn_body
+
+  # Substitute expr for later evaluation
+  if (!is.call(expr)){
+    expr <- substitute(expr)
```
Comment (wch) [inline R/test-module.R]: Isn't this already handled by `testModule`?

File: R/test-module.R [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+  }
+
+  # Create a mock session
+  session <- MockShinySession$new()
+
+  # Parse the additional arguments
+  args <- list(...)
```
Comment (wch) [inline R/test-module.R]: These lines can be consolidated with:

```R
args <- list(..., input = session$input, output = session$output, session = session)
```

File: R/test-module.R (hunk 3 above)
Comment (trestletech) [reply R/test-module.R]: Good catch! Dead code that I forgot to pull out.

File: DESCRIPTION [hunk 5]
```diff
@@ -89,7 +90,9 @@ Suggests:
     ggplot2,
     reactlog (>= 1.0.0),
     magrittr,
-    yaml
+    yaml,
+    future,
```
Comment (wch) [inline DESCRIPTION]: These packages aren't needed anymore, right?

Comment (trestletech) [reply DESCRIPTION]: They're still used in `test-test-module`, actually.
